### DISCUSSION TITLE
Split update for one package by file name

### DISCUSSION THREAD

**Discussion by morremeyer** [renovatebot/renovate#12331] [Request Help] [answered]
I have a use case where I want to split updates for one specific package by the file it is configured in.

Example configuration is laid out in https://github.com/morremeyer/renovate-test to check, but explained here in detail.

Background: There are multiple terraform cloud workspaces provisioned, as follows:

```hcl
resource "tfe_workspace" "awesome" {
  name              = "awesome"
  organization      = "morremeyer"

  # renovate
  terraform_version = "1.0.6"
}
```

The `terraform_version` in there is currently upgraded by the `regexManager`:

```json
{
  "regexManagers": [
    {
      "fileMatch": [
        "\\.tf$"
      ],
      "depNameTemplate": "hashicorp/terraform",
      "datasourceTemplate": "github-tags",
      "extractVersionTemplate": "v(?<version>.*)$",
      "matchStrings": [
        "renovate\\s+terraform_version\\s+=\\s\"(?<currentValue>.*)\""
      ]
    }
  ]
}
```

Each workspace then has a corresponding module, the `awesome` workspace configured above e.g. has the following `backend.tf`:

```hcl
terraform {
  required_version = "1.0.3"
}
```

This is repeated for every provisioned workspace and its corresponding module. To seperate the upgrades of the workspaces from the upgrades of the modules (the workspace needs to be upgraded to the new version first so that a `terraform plan` with the upgraded module version runs against the correct terraform version), the following config is currently in place:

```json
 {
   "packageRules": [
    {
      "matchPackageNames": "hashicorp/terraform",
      "matchManagers": "regex",
      "groupName": "hashicorp-terraform-cloud-workspace",
      "commitMessageSuffix": "for terraform cloud"
    }
  ]
}
```

This separates the terraform version upgrade by the manager so that the workspaces are upgraded in a separate pull request.

When renovate now detects an update for terraform, it will open two pull requests: One for the workspace version, one for the `required_version` in the module.

However, merging the upgrade of the workspace version would block all terraform runs on all workspaces until the `required_version` update for the modules has been merged. To merge the `required_version`, we’ll first have to verify that no unwanted changes occur and fix any possible breakage. 

As this can be complicated and I manage over 30 workspaces currently, I can’t block all terraform runs by merging the workspace upgrade for all workspaces. Therefore, I’d want to separate it by the file they’re in, so that for each file in the [workspaces directory](https://github.com/morremeyer/renovate-test/tree/8e5c4b1fa31d5bf12ee96de45aea2c77664455f1/workspaces), a new PR is opened and for each module (each is in its own directory), a new PR is opened.

Is there any way to achieve this in a sensible configuration?

**Comment by rarkins** [comment, marked as answer]
To split an existing PR into multiple parts essentially means that you need a different branch name. The most elegant way to do this is with `additionalBranchPrefix`, potentially making use of `parentDir` or `baseDir` within package rules. Should be some examples in these search results: https://github.com/renovatebot/renovate/search?q=additionalBranchPrefix+parentDir&type=discussions

**Reply by morremeyer** [reply]
Thank you, that was exactly what I was looking for! For anyone else looking for this, here’s how I configured it in the end:

```json
{
  "regexManagers": [
    {
      "fileMatch": [
        "\\.tf$"
      ],
      "depNameTemplate": "hashicorp/terraform",
      "datasourceTemplate": "github-releases",
      "extractVersionTemplate": "v(?<version>.*)$",
      "matchStrings": [
        "renovate\\s+terraform_version\\s+=\\s\"(?<currentValue>.*)\""
      ]
    }
  ],
  "packageRules": [
    {
      "description": "Separate terraform cloud workspace version updates by file to not block all workspaces at once",
      "matchPackageNames": "hashicorp/terraform",
      "matchManagers": "regex",
      "additionalBranchPrefix": "{{ packageFile }}-",
      "commitMessageSuffix": "for terraform cloud"
    },
    {
      "description": "Separate terraform version updates by module to enable upgrading them one by one",
      "matchPackageNames": "hashicorp/terraform",
      "matchManagers": "terraform",
      "additionalBranchPrefix": "{{ parentDir }}-"
    }
  ]
}
```

This works great as in our case, the directory structure look like this:

```
|
|- workspaces
    |- workspace1.tf
    |- workspace2.tf
|- module-for-workspace-1
|- module-for-workspace-2
```

**Reply by viceice** [reply]
you should use `github-releases` instead of  `github-tags` for terraform 😉

**Reply by morremeyer** [reply]
Fixed, thanks!
